### ISSUE TITLE
Introduce a "getStateObject" in every reducers

### ISSUE THREAD

**Issue by Rewieer** [erikras/ducks-modular-redux#42]
I'm starting to work on my first real Redux project using Ducks and somehow I feel like letting a configureStore function create the whole store and give the name they want is disturbing, because when you want to work on a specific reducer, you have to remember what's the whole state's shape. 

It may become difficult to remember everything, and if we can just find a way to just watch the name somewhere, it could make us earn a lot of time.

So I was thinking about making a "getStateObject" being exported in every reducers. It would look like this : 

`const getStateTree = (state) => state.users;
`
I'm not quite sure this is the place to discuss this, but since this repository is about structuring, why not ?

**Comment by smeijer** [comment]
But it's not the duck that defines on what part of the state the reducer is mounted. It's the `rootReducer` that defines this part. How do you handle this?

```js
import { default as widgets } from './my-widgets/widget-duck';
import { default as otherWidgets } from './other-widgets/widget-duck';

combineReducers({
  otherWidgets,
});
```

I do something similar in my selectors, but I guess it can give some trouble when using it as convention.

```js
import { createSelector } from 'reselect';

const getScope = state => state.users;

export const getCurrentUser = createSelector(getScope,
  scope => scope.get('currentUser'));
```

**Comment by Rewieer** [comment]
After some time working with Redux it comes clear that i'd rather use helpers to retrieve data from store instead of directly accessing state from outside. Something like `getStateTree` doesn't look this useful when thinking this way.

In this idea I always export an object `helpers`, called in my mapStateToProps.

**Comment by elyobo** [comment]
Yes, I have my ducks export helper methods to work with the state to encapsulate the structure of the store within the duck.  Users of the state shouldn't know much about the structure of the store, they just get it and use a helper method to retrieve the data that they need; that way I can change the structure as needed and change the helpers and I don't need to change all the code that accesses that data.

**Comment by Rewieer** [comment]
It seems the correct word for this is "selectors".

**Comment by cdelaorden** [comment]
Hey @Rewieer we were discussing this issue today as well among colleagues, and the problem is not solved by simply exporting selectors from *inside* the module. 

Why? Because from the module's POV all state is simply its state. However, the external reducer combination (either in modules/index or in the typical configureStore/createStore utility) is the one who defines where in the real Redux state tree will your module data will live.

You have either two options: call your selector from your components with the hardcoded "shape" (eg: getMyModuleStuff(state.mymodule)) or, as Dan Abramov does in its second Egghead video series, manually wrap your module selectors with app-wide selectors that select the state slice your module is using.

I think it's a pain for a big application and probably the solution would be to *combine modules* and not only *reducers*, so any selectors/getters are also automatically wrapped with the correct state shape.

Hope this made sense, we'll probably end up implementing this function ourselves. I just thought it would be valuable in this thread in case anyone else is having the same philosophical questions. :)

**Comment by elyobo** [comment]
I take the `getMyModuleStuff(state.mymodule)` approach.  The module itself only knows about the shape of its bit of the state and so that's what it should be given.

**Comment by smeijer** [comment]
@elyobo, makes sense. But at the same time, the components that use the state, often life within the same module.

**Comment by erikras** [comment]
I guess the best argument against this proposal is that, by only exporting selectors, all the code that needs to know about the shape of your state for that reducer is contained in the one file, rather than spread over a bunch of files all over your project that are doing `getStateTree().some.specific.path` to read the data.
